### ISSUE TITLE
Features to 0.3 release

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#420]
## 🚀 Feature
This issue has some features moved over from our `0.2` release #301. They are considered more advanced topics that we'll likely be aiming to have explored or supported for our `0.3` release, which might be around Jun 1 2020.

**Dims features**:
- [ ] Enable range mode for dims including slice / max_projection / crop (S/P/C)
- [x] Add play button to dimension sliders.
- [ ] Support units in status update and scaler bar overlays
- [x] Add axes labels
- [x] Add axes shapes

**File I/O**:
- [x] Loading / saving shapes, points, vectors
- [x] Loading / saving surface
- [x] Loading custom image file types
- [x] Loading / saving entire viewer state to zarr

**Rendering**
- [ ] 3D rendering of big data + image pyramids
- [ ] 3D interactivity / drawing
- [ ] Support mixed 2D and 3D rendering
- [ ] Editing in labels pyramid
- [ ] Lazy rendering

**1D plots**
- [ ] Scatter, line, histogram plots
- [ ] Live updating, with links

**Colors / clims**:
- [ ] For image layer add a clims histogram 
- [ ] Add a colorbar

**Animations**:
- [x] Support for making animations.

**Layers**:
- [ ] Support layer borders #298
- [ ] Support draggable / resizable layers #299
- [x] Add text layer #593 
- [ ] add events on finishing adding shapes / markers / labels / refresh etc

**Multicanvas**:
- [ ] Support multiple canvas with linked cameras
- [ ] For grid mode + coupled zoom
- [ ] Minimap when highly zoomed in
- [ ] For XY, XZ, YZ + 3D views for people doing volumetric imaging

**Provenance logging**
- [ ] Undo / redo buttons
- [ ] History feature
- [ ] Macro generation

**Qt / stylings**
- [ ] autogenerate `_qt_properties` /  `_qt_controls` from function annotations
- [x] display loading icon during start-up
- [x] Make side panels draggable, dockable
- [ ] Support title bar (with editable title and proper stylings)
- [ ] Status bar shows values from multiple layers
- [ ] Add edit boxes next to sliders to set min / max

**Typing / schema**
- [ ] Add typing system for images / shapes
- [ ] Adopt a metadata schema compatible with xarray for images / shapes

**Remote access**
- [ ] remote notebooks
- [ ] web viewing
- [ ] remote data with dask
- [ ] ability to launch from jupyterhub with jupyterlab

**Standalone app**
- [ ] Installable app made with PyInstaller

**Testing**
- [x] add performance benchmarks
- [ ] add better UI tests

**Tutorials and docs**
- [ ] better demo datasets with pooch
- [ ] more application tutorials

**Comment by AhmetCanSolak** [comment]
Hi @sofroniewn , how are you feeling about moving `add play button to dimension sliders` to 0.2.0 milestone? I can work on it after #520 and #496.

**Comment by GenevieveBuckley** [comment]
Does "Support mixed 2D and 3D rendering" have an issue I can follow?

I often have a 3D volume (fluorescence microscopy), and a 2D image (scanning electron microscopy) from the same sample. It'd be really awesome if I could overlay the two.

They generally have very different angles of view from the detector to the sample, as well as very different pixel sizes which makes things kinda tricky.

**Comment by jni** [comment]
> Does "Support mixed 2D and 3D rendering" have an issue I can follow?

A quick search didn't turn up any, so feel free to create one. @sofroniewn has already done some of the leg-work here by allowing layers to have their own dims (dimensions) model, separate from the global one, so I think this one is less of an implementation and more of a UX challenge.

Layers can also have their own scaling and translation. In terms of e.g. 3D rotations or skew, I don't think we have a great story there yet, but I'm not sure.

If you could provide a small sample dataset, that would help enormously with development.

**Comment by sofroniewn** [comment]
We're well underway moving towards these goals, and closing out some of the remaining issues from #301. I just wanted to highlight a couple key areas of work that we need to get done before we get close to having API stability, and having a solid foundation for plugins:

- [ ] Use of colors / colormaps on models #782, #862  

- [ ]  Points vs Text vs Annotation refactor and propagation of changes to Shapes, Vectors #858 

- [ ] 1D plot widgets + live updating #823 

- [ ] More interactivity on the base layer #772 

- [ ] Physical coordinates #763 

- [ ] Camera state / viewer serialization #852 and #851 

- [ ] Finalizing keybindings, mouse bindings, event systems #791 #793

- [ ] Crop / project / slice mode #848, #7 

- [ ] Multicanvas support (linked ortho views, linked grid view) #760 

- [x] FileIO for all layer types #379 

- [ ] Improve remote data support, remote dask server (?) #845 #654

**Comment by liu-ziyang** [comment]
going to close this issue now since we have past 0.3 release cycles, future features will be tracked in 0.4+ release notes
